src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -265,14 +257,107 @@ public NodeStateTable<T> ToImmutableAndFree()
                     return NodeStateTable<T>.Empty;
                 }
 
-                var hasNonCached = _states.Any(static s => s.Any(static i => i.Item2 != EntryState.Cached));
+                var hasNonCached = _states.Any(static s => !s.IsCached);
                 return new NodeStateTable<T>(_states.ToImmutableAndFree(), isCompacted: !hasNonCached);
-
             }
 
             internal ImmutableArray<T> GetLastEntries()
             {
-                return _states[_states.Count - 1].SelectAsArray(t => t.Item1);
+                Debug.Assert(_states.Count > 0);
+                return _states[_states.Count - 1].GetItems();
+            }
+        }
+
+        private readonly struct TableEntry
+        {
+            private static readonly ImmutableArray<EntryState> s_allAddedEntries = ImmutableArray.Create(EntryState.Added);
+            private static readonly ImmutableArray<EntryState> s_allCachedEntries = ImmutableArray.Create(EntryState.Cached);
+            private static readonly ImmutableArray<EntryState> s_allModifiedEntries = ImmutableArray.Create(EntryState.Modified);
+            private static readonly ImmutableArray<EntryState> s_allRemovedEntries = ImmutableArray.Create(EntryState.Removed);
+
+            private readonly ImmutableArray<T> _items;
+            private readonly T? _item;
+            private readonly ImmutableArray<EntryState> _states;
```

Comment:
Can we add a doc comment here? Essentially stating that in the case `_states.Length == 1` that it represent the state for all items and that this is a space saving trick. Also that there will be no case where `_states.Length > 1` and all of the items are the same. Essentially it's an invariant that if all entries in the array are the same then the length of the array is 1.